File: editors/zed/grammars/ripple/grammar.js
/**
 * @file grammar.js — Tree-sitter grammar for the Ripple programming language.
 */

module.exports = grammar({
  name: "ripple",

  word: ($) => $.identifier,

  extras: ($) => [/\s+/],

  conflicts: ($) => [[$.entry, $.assignment_statement]],

  rules: {
    // ── Top-level ─────────────────────────────────────────────────────────────

    source_file: ($) => repeat($.section),

    // ── Section ───────────────────────────────────────────────────────────────

    section: ($) =>
      seq(
        field("header", $.section_header),
        repeat(choice($.entry, $.statement)),
      ),

    /** "[" identifier+ "]"  —  e.g.  [circle of friends], [albums], [jam] */
    section_header: ($) => seq("[", repeat1($.identifier), "]"),

    // ── Entries ───────────────────────────────────────────────────────────────

    entry: ($) =>
      seq(
        field("key", $.identifier),
        ":",
        field("value", choice($.string_literal, $.collection)),
      ),

    collection: ($) => seq("{", repeat(field("item", $.string_literal)), "}"),

    // ── Statements ────────────────────────────────────────────────────────────

    statement: ($) =>
      choice(
        $.assignment_statement,
        $.conditional_statement,
        $.say_statement,
        $.snitch_statement,
      ),

    assignment_statement: ($) =>
      seq(
        field("target", $.identifier),
        "becomes",
        field("value", $.expression),
      ),

    conditional_statement: ($) =>
      seq(
        "suppose",
        field("condition", $.expression),
        repeat(field("body", $.statement)),
        optional(seq("otherwise", repeat(field("alternative", $.statement)))),
        "enough",
      ),

    say_statement: ($) =>
      prec.right(seq("say", repeat1(field("argument", $.expression)))),

    snitch_statement: ($) =>
      prec.right(seq("snitch", repeat1(field("argument", $.expression)))),

    // ── Expressions ───────────────────────────────────────────────────────────

    expression: ($) =>
      seq(
        field("left", $.term),
        optional(seq(field("operator", $.operator), field("right", $.term))),
      ),

    operator: ($) =>
      choice(
        "vibes_like",
        "harshing_the_vibe_of",
        "louder_than",
        "quieter_than",
        "has",
        "and",
        "or",
      ),

    // ── Terms ─────────────────────────────────────────────────────────────────

    term: ($) =>
      choice(
        prec(1, $.attribute),
        $.string_literal,
        $.boolean_literal,
        $.identifier,
      ),

    attribute: ($) =>
      seq(field("object", $.identifier), ".", field("property", $.identifier)),

    boolean_literal: ($) => choice("copacetic", "harsh"),

    // ── Tokens ────────────────────────────────────────────────────────────────

    string_literal: (_) => /"[^"]*"/,

    identifier: (_) => /[a-zA-Z_][a-zA-Z0-9_]*/,
  },
});

File: internal/ast/grammar.go
package ast

import (
	"github.com/alecthomas/participle/v2"
	"github.com/alecthomas/participle/v2/lexer"
)

var ripleLexer = lexer.MustStateful(lexer.Rules{
	"Root": {
		{Name: "Comment", Pattern: `#[^\n\r]*`},
		{Name: "String", Pattern: `"[^"]*"`},
		{Name: "BracketPhrase", Pattern: `\[[a-zA-Z_][a-zA-Z0-9_]*(?:\s+[a-zA-Z_][a-zA-Z0-9_]*)*\]`},
		{Name: "Punct", Pattern: `[[\]{}.:]`},
		{Name: "Keyword", Pattern: `(?:becomes|suppose|say|snitch|otherwise|enough|harshing_the_vibe_of|copacetic|harsh|vibes_like|louder_than|quieter_than|has|and|or)\b`},
		{Name: "Ident", Pattern: `[a-zA-Z_][a-zA-Z0-9_]*`},
		{Name: "Newline", Pattern: `\r?\n`},
		{Name: "Whitespace", Pattern: `[ \t]+`},
	},
})

type Program struct {
	Sections []*Section `Newline* ( @@ Newline* )*`
}

type Section struct {
	Pos     lexer.Position
	Friends *CircleOfFriendsSection `@@`
	Albums  *AlbumsSection          `| @@`
	Jam     *JamSection             `| @@`
}

type JamSection struct {
	Pos        lexer.Position
	Name       JamPhrase    `@BracketPhrase Newline*`
	Statements []*Statement `( @@ Newline* )*`
}

type CircleOfFriendsSection struct {
	Pos     lexer.Position
	Name    FriendsPhrase  `@BracketPhrase Newline*`
	Entries []*FriendEntry `( @@ Newline* )*`
}

type AlbumsSection struct {
	Pos     lexer.Position
	Name    AlbumsPhrase  `@BracketPhrase Newline*`
	Entries []*AlbumEntry `( @@ Newline* )*`
}

type FriendEntry struct {
	Pos   lexer.Position
	Key   string `@Ident ":"`
	Value string `@String`
}

type AlbumEntry struct {
	Pos        lexer.Position
	Key        string   `@Ident ":"`
	Collection []string `"{" Newline* @String* (Newline* @String)* Newline* "}"`
}

type Statement struct {
	Pos         lexer.Position
	Assignment  *Assignment  `  @@`
	Conditional *Conditional `| @@`
	Print       *Print       `| @@`
	Snitch      *Snitch      `| @@`
}

type Assignment struct {
	Pos    lexer.Position
	Target string      `@Ident "becomes"`
	Value  *Expression `@@`
}

type Snitch struct {
	Pos  lexer.Position
	Args []*Expression `"snitch" @@+`
}

type Print struct {
	Pos  lexer.Position
	Args []*Expression `"say" @@+`
}

type Conditional struct {
	Pos       lexer.Position
	Condition *Expression  `"suppose" @@ Newline*`
	Body      []*Statement `( @@ Newline* )*`
	Otherwise []*Statement `( "otherwise" Newline* ( @@ Newline* )* )?`
	End       string       `"enough"`
}

type Expression struct {
	Pos   lexer.Position
	Left  *Term  `@@`
	Op    string `[ @( "louder_than" | "quieter_than" | "vibes_like" | "harshing_the_vibe_of" | "has" | "and" | "or" )`
	Right *Term  `  @@ ]`
}

type Term struct {
	Pos   lexer.Position
	Attr  *Attr   `  @@`
	Str   *string `| @String`
	Bool  *string `| @( "copacetic" | "harsh" )`
	Ident *string `| @Ident`
}

type Attr struct {
	Pos  lexer.Position
	Name string `@Ident "."`
	Prop string `@Ident`
}

var Parser = participle.MustBuild[Program](
	participle.Lexer(ripleLexer),
	participle.Unquote("String"),
	participle.UseLookahead(5),
	participle.Elide("Whitespace", "Comment"),
)
